torch/utils/mem_tracker.py

```diff
@@ -0,0 +1,510 @@
+            print(f"Module:  {mod_fqn}")
+            for state, snapshot in mod_stats.snapshots.items():
+                print(f"{state}")
+                self.print_mem_snapshot(False, units, snapshot)
+            print()
+
+    def _track_module_params_and_buffers(
+        self, module: nn.Module, install_grad_hooks: bool = True
+    ) -> Tuple[int, int]:
+        def _grad_hook(param: nn.Parameter):
+            if param.grad is not None:
+                WeakRefInfo.update_and_maybe_create_winfo(
+                    param.grad,
+                    MemRefType.gradient,
+                    self.WINFO,
+                    update_existing=True,
+                )
+
+        param_memory = 0
+        for param in module.parameters():
+            winfo = WeakRefInfo.update_and_maybe_create_winfo(
+                param, MemRefType.parameter, self.WINFO
+            )
+            param_memory += winfo.mem_consumed
+            if param.grad is not None:
+                WeakRefInfo.update_and_maybe_create_winfo(
+                    param.grad, MemRefType.gradient, self.WINFO
+                )
+            if (
+                self._param_to_grad_hook_handles.get(param, None) is None
+                and install_grad_hooks
+            ):
+                grad_hook_handle = param.register_post_accumulate_grad_hook(_grad_hook)
+                self._param_to_grad_hook_handles[param] = grad_hook_handle
+        buffer_memory = 0
+        for buffer in module.buffers():
+            winfo = WeakRefInfo.update_and_maybe_create_winfo(
+                buffer, MemRefType.buffer, self.WINFO
+            )
+            buffer_memory += winfo.mem_consumed
+        return (param_memory, buffer_memory)
+
+    def _track_inputs_or_outputs(self, args: Any) -> int:
+        input_or_output_memory = 0
+
+        def add_inps_or_outs(t: torch.Tensor):
+            nonlocal input_or_output_memory
+            st = WeakRefInfo.get_untyped_storage(t)
+            if winfo := WeakRefInfo.get_winfo_if_exists(st, self.WINFO):
+                input_or_output_memory += winfo.mem_consumed
+
+        tree_map_only(torch.Tensor, add_inps_or_outs, args)
+        return input_or_output_memory
+
+    def _pre_fw_hook(self, module: nn.Module, inputs: Any):
+        mod_name = self.mod_tracker.get_known_fqn(module)
+        assert mod_name is not None and mod_name in self.mod_tracker.parents
+
+        if module not in self.memory_tracking:
+            mod_stats = ModMemStats(mod_name)
+            self.memory_tracking[module] = mod_stats
+            state = ModState.pre_fw
+            param_mem, buffer_mem = self._track_module_params_and_buffers(
+                module, install_grad_hooks=True
+            )
+            input_mem = self._track_inputs_or_outputs(inputs)
+            mod_stats.parameter_mem = param_mem
+            mod_stats.buffer_mem = buffer_mem
+            mod_stats.input_mem = input_mem
+
+        elif self.mod_tracker.is_bw:
```

Comment:
If it's the first time we see a module in pre-fw we create an entry for it in memory_tracking. If we encounter pre-fw in backward it is for AC, else it is the second pre-fw call so we override the stats, which I now think is wrong since there could be multiple fw calls for the same module.